algebra/free_group.lean

```diff
@@ -0,0 +1,423 @@
+| mul_left_inv : ∀ x, rel (x⁻¹ * x) 1
+
+instance quotient_rel : setoid M :=
+⟨rel M, rel.refl, rel.symm, rel.trans⟩
+
+end inv_mon.to_group
+
+@[reducible] def inv_mon.to_group (M : Type u) [inv_mon M] : Type u :=
+quotient (inv_mon.to_group.quotient_rel M)
+
+namespace inv_mon.to_group
+
+instance (M : Type u) [inv_mon M] : group (inv_mon.to_group M) :=
+{ mul := λ x y, quotient.lift_on₂ x y (λ m n, quotient.mk (m * n))
+    (λ x y z w hxz hyw, quotient.sound $ rel.mul x y z w hxz hyw),
+  mul_assoc := λ x y z, quotient.induction_on₃ x y z $
+    λ a b c, quotient.sound $ by rw mul_assoc,
+  one := quotient.mk 1,
+  mul_one := λ x, quotient.induction_on x $
+    λ a, quotient.sound $ by rw mul_one,
+  one_mul := λ x, quotient.induction_on x $
+    λ a, quotient.sound $ by rw one_mul,
+  inv := λ x, quotient.lift_on x (λ m, quotient.mk m⁻¹)
+    (λ x y hxy, quotient.sound $ rel.inv x y hxy),
+  mul_left_inv := λ x, quotient.induction_on x $
+    λ a, quotient.sound $ rel.mul_left_inv a }
+
+def of_inv_mon {M : Type u} [inv_mon M] : M → inv_mon.to_group M :=
+quotient.mk
+
+@[simp] lemma of_inv_mon.mul {M : Type u} [inv_mon M] {x y : M} :
+  of_inv_mon (x * y) = of_inv_mon x * of_inv_mon y := rfl
+
+@[simp] lemma of_inv_mon.one {M : Type u} [inv_mon M] :
+  of_inv_mon (1:M) = 1 := rfl
+
+@[simp] lemma of_inv_mon.inv {M : Type u} [inv_mon M] {x : M} :
+  of_inv_mon x⁻¹ = (of_inv_mon x)⁻¹ := rfl
+
+section left_adjoint
+
+parameters (M : Type u) [inv_mon M]
+parameters (G : Type v) [group G]
+parameters (f : M → G) -- M → Forgetful(G)
+parameters (Hmul : ∀ x y, f (x * y) = f x * f y)
+parameters (Hone : f 1 = 1)
+parameters (Hinv : ∀ x, f x⁻¹ = (f x)⁻¹)
+include Hmul Hone Hinv
+
+def left_adjoint : inv_mon.to_group M → G := -- Free(M) → G
+λ x, quotient.lift_on x f $ λ a b hab, begin
+  induction hab with h c d h ih c d e h1 h2 ih1 ih2
+    c d p q h1 h2 ih1 ih2 c d h ih c,
+  case inv_mon.to_group.rel.refl
+  { refl },
+  case inv_mon.to_group.rel.symm
+  { exact ih.symm },
+  case inv_mon.to_group.rel.trans
+  { exact ih1.trans ih2 },
+  case inv_mon.to_group.rel.mul
+  { rw [Hmul, Hmul, ih1, ih2] },
+  case inv_mon.to_group.rel.inv
+  { rw [Hinv, Hinv, ih] },
+  case inv_mon.to_group.rel.mul_left_inv
+  { rw [Hmul, Hinv, mul_left_inv, Hone] }
+end
+
+theorem left_adjoint.is_group_hom : is_group_hom left_adjoint :=
+λ x y, quotient.induction_on₂ x y Hmul
+
+theorem left_adjoint.commutes (x) : left_adjoint (of_inv_mon x) = f x := rfl
+
+parameters (g : inv_mon.to_group M → G)
+parameters (Hg : ∀ x, g (of_inv_mon x) = f x)
+
+theorem left_adjoint.unique : ∀ x, g x = left_adjoint x :=
+λ x, quotient.induction_on x $ λ m, Hg m
+
+end left_adjoint
+
+end inv_mon.to_group
+
+class inv_type (IT : Type u) extends has_inv IT :=
+(inv_inv : ∀ x : IT, x⁻¹⁻¹ = x)
+
+instance inv_mon.to_inv_type (M : Type u) [inv_mon M] : inv_type M :=
+{ inv_inv := inv_mon.inv_inv }
+
+@[reducible] def inv_type.to_inv_mon (IT : Type u) [inv_type IT] :=
```

Comment:
I would not introduce an extra name for it. Is there a specific reason why you don't want to use `list`?